src/compute/src/bin/computed.rs

```diff
@@ -280,52 +273,33 @@ async fn run(args: Args) -> Result<(), anyhow::Error> {
         client = Box::new(ComputeCommandReconcile::new(client))
     }
 
-    let mut _pid_file = None;
-    if let Some(pid_file_location) = &args.pid_file_location {
-        _pid_file = Some(PidFile::open(&pid_file_location).unwrap());
-    }
-
     serve(serve_config, server, client).await
 }
 
 struct ServeConfig {
-    listener: TcpListener,
+    listen_addr: String,
     linger: bool,
 }
 
-async fn serve<G, C, R>(
-    config: ServeConfig,
-    _server: Server,
-    mut client: G,
-) -> Result<(), anyhow::Error>
+async fn serve<G>(config: ServeConfig, _server: Server, mut client: G) -> Result<(), anyhow::Error>
 where
-    G: GenericClient<C, R>,
-    C: DeserializeOwned + fmt::Debug + Send + Unpin,
-    R: Serialize + fmt::Debug + Send + Unpin,
+    G: GenericClient<ComputeCommand, ComputeResponse>,
 {
-    loop {
-        let (conn, _addr) = config.listener.accept().await?;
-        info!("coordinator connection accepted");
+    let (mut grpc_serve, mut shutdown_rx) =
+        mz_dataflow_types::client::tcp::grpc_computed_server(config.listen_addr, config.linger);
 
-        let mut conn = mz_dataflow_types::client::tcp::framed_server(conn);
-        loop {
-            select! {
-                cmd = conn.try_next() => match cmd? {
-                    None => break,
-                    Some(cmd) => { client.send(cmd).await.unwrap(); },
-                },
-                res = client.recv() => {
-                    match res.unwrap() {
-                        None => break,
-                        Some(response) => { conn.send(response).await?; }
-                    }
+    loop {
+        select! {
+            cmd = grpc_serve.recv() => {
+                client.send(cmd?).await.unwrap()
+            },
+            res = client.recv() => {
+                match res.unwrap() {
+                    None => { },
+                    Some(response) => grpc_serve.send(response).await
                 }
-            }
-        }
-        if !config.linger {
-            break;
-        } else {
-            info!("coordinator connection gone; lingering");
+            },
+            _ = shutdown_rx.recv() => break
```

Comment:
We should check that each of the `recv` calls is cancellation-safe.

Reply:
Indeed. So grpc recv() has three await points
1) awaiting the mutex to access the `rx_queue`
2) awaiting a message on this `rx_queue`
3) in case there is no queue (yet), it awaits on a notification.

Number 1) is idempotent, 2) is not, but if we pass this await point we have received a message, which in turn we immediately reply (ie. no further awaits). 3) is a barrier, but it's not possible to miss the notification in a cancellation as the retry would check the `shared.queue` first.

There is also no race between the creation of new queue pair, as the Notify will store internally a permit: If a notifier comes first, and than notified awaited, it will immediately return (and clear the permit). See https://docs.rs/tokio/0.2.12/tokio/sync/struct.Notify.html

We still have to think about ComputeClient (and all it's possible instantiations with reconcile etc.).

Reply:
Can you add this justification as a comment?

we may end up just rewriting all the `select!` code without `select!` instead of justifying all the `ComputeClient` usages